src/nvim/lua/vim.lua

```diff
@@ -489,4 +489,73 @@ function vim.defer_fn(fn, timeout)
   return timer
 end
 
+local on_keystroke_callbacks = {}
+
+--- Register a lua {fn} with an {id} to be run after every keystroke.
+---
+--- NOTE: {fn} will be automatically removed if an error occurs while calling.
+---
+--@param ns_id number: See |nvim_create_namesapce()|
+--@param fn function: Function to call. It should take one argument, which is a string.
+--                    The string will contain the literal keys typed.
+--                    See |i_CTRL-V|
+function vim.register_keystroke_callback(ns_id, fn)
+  vim.validate {
+    fn = { fn, 'c', true},
+    ns_id = { ns_id, function()
+      if type(ns_id) ~= 'number' then
+        return false
+      end
+
+      return ns_id > 0
+    end, 'Must be valid namespace ID. See |:help nvim_create_namespace()|' }
+  }
+
+  on_keystroke_callbacks[ns_id] = fn
+  return ns_id
+end
+
+--- Remove {id} from running on keystroke.
+---
+--@param ns_id number: See |nvim_create_namesapce()|
+function vim.remove_keystroke_callback(ns_id)
+  vim.validate { ns_id = { ns_id, 'n' } }
+
+  on_keystroke_callbacks[ns_id] = nil
+end
+
+--- Function that executes the keystroke callbacks.
+--@private
+function vim._log_keystroke(char)
```

Comment:
Alternatively
```lua
function vim._log_keystroke(char)
  local failed_indices = {}
  local failed_messages = {}
  for k, v in pairs(on_keystroke_callbacks) do
    local ok, err_msg = pcall(v, char)
    if not ok then
      on_keystroke_callbacks[k] = nil
      table.insert(failed_indices, k)
      table.insert(failed_messages, err_msg)
    end
  end

  if failed_indices[1] then
    error(string.format(
      "Error executing 'on_keystroke' with indices of '%s'\n    With messages: %s",
      table.concat(failed_indices, ", "),
      table.concat(failed_messages, "\n")))
  end
end
```
Note that deleting from a table is allowed during iteration with pairs. Quote:

> The behavior of next is undefined if, during the traversal, you assign any value to a non-existent field in the table. You may however modify existing fields. In particular, you may clear existing fields.